File: killer_sudoku/solver/engine/work_queue.py
"""WorkItem and SolverQueue — priority queue with deduplication and version tracking.

WorkItem has three forms:
- Unit-scoped (all triggers except CELL_DETERMINED, CELL_SOLVED, and GLOBAL):
    carries unit_id and unit_version for staleness detection
- Cell-scoped (CELL_DETERMINED and CELL_SOLVED):
    carries cell; never version-skipped
- GLOBAL (no unit or cell):
    never version-skipped

Deduplication keys:
- Unit-scoped: (id(rule), unit_id)  — one item per rule/unit pair
- Cell-scoped: (id(rule), cell)     — one item per rule/cell pair
- GLOBAL:      (id(rule),)           — one item per GLOBAL rule
"""

from __future__ import annotations

import dataclasses
import heapq
from typing import Any

from killer_sudoku.solver.engine.rule import SolverRule
from killer_sudoku.solver.engine.types import Cell, Trigger


@dataclasses.dataclass
class WorkItem:
    """A single unit of work for the solver engine main loop."""

    priority: int
    rule: SolverRule
    unit_id: int | None
    unit_version: int | None  # snapshot of board.unit_versions[unit_id] at enqueue
    cell: Cell | None
    trigger: Trigger
    hint_digit: int | None

    def dedup_key(self) -> tuple[Any, ...]:
        """Key used to identify duplicate items in the queue."""
        if self.trigger in (Trigger.CELL_DETERMINED, Trigger.CELL_SOLVED):
            return (id(self.rule), self.cell)
        if self.trigger == Trigger.GLOBAL:
            return (id(self.rule),)
        return (id(self.rule), self.unit_id)

    def is_stale(self, unit_versions: list[int]) -> bool:
        """True if no candidate was removed in this unit since enqueue.

        Cell-scoped (CELL_DETERMINED, CELL_SOLVED) and GLOBAL items are never
        stale — they represent a specific event (not a unit state) and must
        always be processed.
        """
        cell_scoped = (Trigger.CELL_DETERMINED, Trigger.CELL_SOLVED)
        if self.trigger in cell_scoped or self.trigger == Trigger.GLOBAL:
            return False
        if self.unit_id is None or self.unit_version is None:
            return False
        return unit_versions[self.unit_id] == self.unit_version

    def __lt__(self, other: WorkItem) -> bool:
        return (self.priority, self.rule.name) < (other.priority, other.rule.name)


class SolverQueue:
    """Min-heap priority queue with O(1) deduplication.

    When the same (rule, unit_id) pair is enqueued twice, the item with the
    lower priority wins; the newer trigger/hint replaces the older one.
    Superseded entries remain in the heap as ghosts and are discarded on pop.
    """

    def __init__(self) -> None:
        self._heap: list[WorkItem] = []
        # dedup_key -> best priority currently live in the queue
        self._best: dict[tuple[Any, ...], int] = {}

    def enqueue_unit(
        self,
        priority: int,
        rule: SolverRule,
        unit_id: int,
        unit_version: int,
        trigger: Trigger,
        hint_digit: int | None,
    ) -> None:
        """Enqueue a unit-scoped work item with deduplication."""
        key = (id(rule), unit_id)
        existing = self._best.get(key)
        if existing is not None and existing <= priority:
            return  # already have a better or equal item
        self._best[key] = priority
        heapq.heappush(
            self._heap,
            WorkItem(priority, rule, unit_id, unit_version, None, trigger, hint_digit),
        )

    def enqueue_cell(
        self,
        priority: int,
        rule: SolverRule,
        cell: Cell,
        trigger: Trigger,
        hint_digit: int | None,
    ) -> None:
        """Enqueue a cell-scoped work item with deduplication."""
        key = (id(rule), cell)
        if key in self._best:
            return  # already queued for this cell
        self._best[key] = priority
        heapq.heappush(
            self._heap,
            WorkItem(priority, rule, None, None, cell, trigger, hint_digit),
        )

    def enqueue_global(self, priority: int, rule: SolverRule) -> None:
        """Enqueue a GLOBAL work item — at most one per rule at a time."""
        key = (id(rule),)
        if key in self._best:
            return
        self._best[key] = priority
        heapq.heappush(
            self._heap,
            WorkItem(priority, rule, None, None, None, Trigger.GLOBAL, None),
        )

    def pop(self) -> WorkItem:
        """Pop the lowest-priority live item. Discards superseded ghosts."""
        while self._heap:
            item = heapq.heappop(self._heap)
            key = item.dedup_key()
            if self._best.get(key) == item.priority:
                del self._best[key]
                return item
        raise IndexError("pop from empty SolverQueue")

    def empty(self) -> bool:
        """True if no live items remain in the queue."""
        return len(self._best) == 0

File: web/src/engine/workQueue.ts
/**
 * WorkItem and SolverQueue — priority queue with deduplication and version tracking.
 *
 * Mirrors Python's `killer_sudoku.solver.engine.work_queue` module.
 *
 * WorkItem has three forms:
 *  - Unit-scoped: carries unitId and unitVersion for staleness detection.
 *  - Cell-scoped (CELL_DETERMINED, CELL_SOLVED): carries cell; never stale.
 *  - GLOBAL: no unit or cell; never stale.
 *
 * Deduplication keys (matching Python id(rule) logic):
 *  - Unit-scoped:  `${ruleIdx}:${unitId}`
 *  - Cell-scoped:  `${ruleIdx}:${cell[0]},${cell[1]}`
 *  - GLOBAL:       `${ruleIdx}`
 */

import type { SolverRule } from './rule.js';
import { Trigger } from './types.js';
import type { Cell } from './types.js';

/** A single unit of work for the solver engine main loop. */
export interface WorkItem {
  readonly priority: number;
  readonly rule: SolverRule;
  readonly ruleIdx: number;       // stable per-rule index used for dedup keys
  readonly unitId: number | null;
  readonly unitVersion: number | null;
  readonly cell: Cell | null;
  readonly trigger: Trigger;
  readonly hintDigit: number | null;
}

function dedupKey(item: WorkItem): string {
  if (item.trigger === Trigger.CELL_DETERMINED || item.trigger === Trigger.CELL_SOLVED) {
    return `${item.ruleIdx}:${item.cell![0]},${item.cell![1]}`;
  }
  if (item.trigger === Trigger.GLOBAL) {
    return `${item.ruleIdx}`;
  }
  return `${item.ruleIdx}:${item.unitId}`;
}

export function isStale(item: WorkItem, unitVersions: number[]): boolean {
  if (
    item.trigger === Trigger.CELL_DETERMINED ||
    item.trigger === Trigger.CELL_SOLVED ||
    item.trigger === Trigger.GLOBAL
  ) return false;
  if (item.unitId === null || item.unitVersion === null) return false;
  return unitVersions[item.unitId] === item.unitVersion;
}

/**
 * Min-heap priority queue with O(1) deduplication.
 *
 * When the same (rule, unitId) pair is enqueued twice, the item with the lower
 * priority wins. Superseded entries remain in the heap as ghosts and are
 * discarded on pop.
 */
export class SolverQueue {
  private _heap: WorkItem[] = [];
  /** dedup key → best priority currently live */
  private _best: Map<string, number> = new Map();

  enqueueUnit(
    priority: number,
    rule: SolverRule,
    ruleIdx: number,
    unitId: number,
    unitVersion: number,
    trigger: Trigger,
    hintDigit: number | null,
  ): void {
    const key = `${ruleIdx}:${unitId}`;
    const existing = this._best.get(key);
    if (existing !== undefined && existing <= priority) return;
    this._best.set(key, priority);
    this._push({ priority, rule, ruleIdx, unitId, unitVersion, cell: null, trigger, hintDigit });
  }

  enqueueCell(
    priority: number,
    rule: SolverRule,
    ruleIdx: number,
    cell: Cell,
    trigger: Trigger,
    hintDigit: number | null,
  ): void {
    const key = `${ruleIdx}:${cell[0]},${cell[1]}`;
    if (this._best.has(key)) return;
    this._best.set(key, priority);
    this._push({ priority, rule, ruleIdx, unitId: null, unitVersion: null, cell, trigger, hintDigit });
  }

  enqueueGlobal(priority: number, rule: SolverRule, ruleIdx: number): void {
    const key = `${ruleIdx}`;
    if (this._best.has(key)) return;
    this._best.set(key, priority);
    this._push({ priority, rule, ruleIdx, unitId: null, unitVersion: null, cell: null, trigger: Trigger.GLOBAL, hintDigit: null });
  }

  pop(): WorkItem {
    for (;;) {
      if (this._heap.length === 0) throw new Error('pop from empty SolverQueue');
      const item = this._popHeap()!;
      const key = dedupKey(item);
      if (this._best.get(key) === item.priority) {
        this._best.delete(key);
        return item;
      }
      // ghost — discard and continue
    }
  }

  empty(): boolean {
    return this._best.size === 0;
  }

  // --- Min-heap operations ---

  private _push(item: WorkItem): void {
    this._heap.push(item);
    this._siftUp(this._heap.length - 1);
  }

  private _popHeap(): WorkItem | undefined {
    const heap = this._heap;
    if (heap.length === 0) return undefined;
    const top = heap[0];
    const last = heap.pop()!;
    if (heap.length > 0) {
      heap[0] = last;
      this._siftDown(0);
    }
    return top;
  }

  private _lt(a: WorkItem, b: WorkItem): boolean {
    if (a.priority !== b.priority) return a.priority < b.priority;
    return a.rule.name < b.rule.name;
  }

  private _siftUp(i: number): void {
    const heap = this._heap;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this._lt(heap[i], heap[parent])) {
        [heap[i], heap[parent]] = [heap[parent], heap[i]];
        i = parent;
      } else break;
    }
  }

  private _siftDown(i: number): void {
    const heap = this._heap;
    const n = heap.length;
    for (;;) {
      let smallest = i;
      const l = 2 * i + 1, r = 2 * i + 2;
      if (l < n && this._lt(heap[l], heap[smallest])) smallest = l;
      if (r < n && this._lt(heap[r], heap[smallest])) smallest = r;
      if (smallest === i) break;
      [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
      i = smallest;
    }
  }
}
